### Домашнее задание

1. Разделить дата сет на трейн и тест в отношение 50:50 70:30 80:20 (с перемешиванием)
2. Обучать наши модели на трейне. Предсказывать и замерять метрику R^2 и на трейне и на тесте
3. Проверить следующие модели, для каждого разделения: а) sales ~ log_tv + radio б) sales ~ TV + radio в) sales ~ TV + radio + newspaper



In [1]:
# Common imports
import pandas as pd
import numpy as np
import statsmodels.api as sm
import statsmodels.formula.api as smf

# to make this notebook's output stable across runs
np.random.seed(42)

# To plot pretty figures
import matplotlib
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import axes3d
%matplotlib inline

plt.rcParams['axes.labelsize'] = 14
plt.rcParams['xtick.labelsize'] = 12
plt.rcParams['ytick.labelsize'] = 12
plt.rcParams['figure.figsize'] = (10, 5)

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
import math


In [2]:
import os
import urllib
import shutil

def download_file(url, dir_path="data"):
    if not os.path.exists(dir_path):
        os.makedirs(dir_path)
        
    file_name = os.path.split(url)[-1]
    file_path = os.path.join(dir_path, file_name)
    
    with urllib.request.urlopen(url) as response, open(file_path, 'wb') as out_file:
        shutil.copyfileobj(response, out_file)
        
    return file_path

In [3]:
# Nado?
from sklearn.preprocessing import StandardScaler, normalize, MinMaxScaler

In [4]:
adv_df = pd.read_csv('data/Advertising.csv', usecols=[1,2,3,4])
adv_df.head()

TV  radio  newspaper  sales
0  230.1   37.8       69.2   22.1
1   44.5   39.3       45.1   10.4
2   17.2   45.9       69.3    9.3
3  151.5   41.3       58.5   18.5
4  180.8   10.8       58.4   12.9

# Начало

In [5]:
# Добавить кол-ку sq_tv
# adv_df['sq_log'] = adv_df.TV.apply(lambda x: math.log(x))
adv_df['sq_tv'] = adv_df.TV.apply(lambda x: math.pow(x, 0.4))

# Разделить дата сет на трейн и тест в отношение 50:50 70:30 80:20 (с перемешиванием
y = adv_df['sales'].copy()
adv_train, adv_test, y_train, y_test   = train_test_split(adv_df, y, test_size=0.5, random_state=42)
adv_train.shape, adv_test.shape

((100, 5), (100, 5))

#### sales ~ TV + radio + newspaper + sq_tv 

In [6]:
# Определяем данные
x_cols = adv_train.columns.drop("sales") 
y_col = 'sales'

x_train = np.array(adv_train[x_cols])
x_test = np.array(adv_test[x_cols])
y_train = adv_train[y_col]
y_test = adv_test[y_col]
                  
# Обучение на train LinearRegression
lm = LinearRegression().fit(
    x_train,
    y_train
)                  
                                    
# train model предсказать
y_train_preds = lm.predict(x_train)
    
# test model предсказать
y_test_preds = lm.predict(x_test) 

минимизируем RSE и максимизируем 𝑅2. 𝑅2∈[0,1] - относительная величина, чем ближе к 1, тем лучше

In [8]:
# Посчитать R^2 train
lm.score(x_train, y_train)

0.9285475356390764

In [9]:
# Посчитать R^2 test 
lm.score(x_test, y_test)

0.9248768400230458

#### sales ~ sq_tv + radio

In [10]:
# Определяем данные
x_cols = ['sq_tv','radio']
y_col = 'sales'

x_train = np.array(adv_train[x_cols])
x_test = np.array(adv_test[x_cols])
y_train = adv_train[y_col]
y_test = adv_test[y_col]
                  
# Обучение на train LinearRegression
lm = LinearRegression().fit(
    x_train,
    y_train
)                  
                                    
# train model предсказать
y_train_preds = lm.predict(x_train)
    
# test model предсказать
y_test_preds = lm.predict(x_test) 

In [11]:
# Посчитать R^2 train
lm.score(x_train, y_train)

0.9273049082016976

In [12]:
# Посчитать R^2 test 
lm.score(x_test, y_test)

0.9310741259261295

#### sales ~ TV + radio

In [13]:
# Определяем данные
x_cols = ['TV','radio']
y_col = 'sales'

x_train = np.array(adv_train[x_cols])
x_test = np.array(adv_test[x_cols])
y_train = adv_train[y_col]
y_test = adv_test[y_col]
                  
# Обучение на train LinearRegression
lm = LinearRegression().fit(
    x_train,
    y_train
)                  
                                    
# train model предсказать
y_train_preds = lm.predict(x_train)
    
# test model предсказать
y_test_preds = lm.predict(x_test) 

In [14]:
# Посчитать R^2 train
lm.score(x_train, y_train)

0.9020506014720118

In [15]:
# Посчитать R^2 test 
lm.score(x_test, y_test)

0.8826436017134701

#### sales ~ TV + radio + newspaper

In [16]:
# Определяем данные
x_cols = ['TV','radio','newspaper']
y_col = 'sales'

x_train = np.array(adv_train[x_cols])
x_test = np.array(adv_test[x_cols])
y_train = adv_train[y_col]
y_test = adv_test[y_col]
                  
# Обучение на train LinearRegression
lm = LinearRegression().fit(
    x_train,
    y_train
)                  
                                    
# train model предсказать
y_train_preds = lm.predict(x_train)
    
# test model предсказать
y_test_preds = lm.predict(x_test) 

In [17]:
# Посчитать R^2 train
lm.score(x_train, y_train)

0.9042613648908894

In [18]:
# Посчитать R^2 test 
lm.score(x_test, y_test)

0.8721004816045136

Для 80:20: везде test значение R^2 выше, чем train.
Наивысшее значени R^2 test = 0.942563909350695 для sales ~ sq_tv + radio. Good choice.

Для 70:30: везде test значение R^2 ниже, чем train.
Наивысшее значени R^2 test = 0.9271446043449548 для sales ~ sq_tv + radio.

Для 50:50: везде test значение R^2 ниже, чем train, кроме sales ~ sq_tv + radio
Наивысшее значени R^2 test = 0.9310741259261295 для sales ~ sq_tv + radio.